### PR TITLE
Inputdriver5: backend code for inputdriver configuration

### PR DESCRIPTION
**Pull request by MichaelPFrey**
* back-end code for setting the joystick number
* spaceNavDominantAxisOnly option

### CODE REVIEW

File: src/input/JoystickInputDriver.cc [hunk 1]
```diff
@@ -93,3 +95,8 @@ const std::string & JoystickInputDriver::get_name() const
     static std::string name = "JoystickInputDriver";
     return name;
 }
+
+void JoystickInputDriver::setJoystickNr(std::string jnr)
+{
+    nr=jnr;
```
Comment (kintel) [inline src/input/JoystickInputDriver.cc]: Prefix member access with `this->` to make it more clear what's being changed.
Also (Out of scope for this PR), more descriptive variable names would make code more readable by future readers.

File: src/input/JoystickInputDriver.h [hunk 2]
```diff
@@ -35,11 +35,13 @@ class JoystickInputDriver : public InputDriver
...
     bool open() override;
     void close() override;
-
     const std::string & get_name() const  override;
+    void setJoystickNr(std::string jnr);
+
 private:
     int fd;
     int version;
+    std::string nr="0";
```
Comment (kintel) [inline src/input/JoystickInputDriver.h]: Prefer uniform initialization:
`std::string nr{"0"};`

File: src/input/SpaceNavInputDriver.h [hunk 3]
```diff
@@ -38,9 +38,11 @@ class SpaceNavInputDriver : public InputDriver
     void run() override;
     bool open() override;
     void close() override;
+    void setDominantAxisOnly(bool var);
 
     const std::string & get_name() const override;
 
 private:
     bool spnav_input();
+    bool DominantAxisOnly=true;
```
Comment (kintel) [inline src/input/SpaceNavInputDriver.h]: Member variables should start with a lower-case character. Also, uniform initialization is preferable:
`bool dominantAxisOnly{true};`

File: src/openscad.cc [hunk 4]
```diff
@@ -791,14 +792,22 @@ int gui(vector<string> &inputFiles, const fs::path &original_path, int argc, cha
     app.connect(&app, SIGNAL(lastWindowClosed()), &app, SLOT(releaseQSettingsCached()));
 	app.connect(&app, SIGNAL(lastWindowClosed()), &app, SLOT(quit()));
 	if (Feature::ExperimentalInputDriver.is_enabled()) {
+		Settings::Settings *s = Settings::Settings::inst();
```
Comment (kintel) [inline src/openscad.cc]: You could `auto` this.

File: src/openscad.cc [hunk 5]
```diff
@@ -791,14 +792,22 @@ int gui(vector<string> &inputFiles, const fs::path &original_path, int argc, cha
...
 #ifdef ENABLE_HIDAPI
         InputDriverManager::instance()->registerDriver(new HidApiInputDriver());
 #endif
 #ifdef ENABLE_SPNAV
-        InputDriverManager::instance()->registerDriver(new SpaceNavInputDriver());
+        SpaceNavInputDriver spaceNav = new SpaceNavInputDriver();
```
Comment (kintel) [inline src/openscad.cc]: This looks odd; should this be a pointer?
It could be fixed by `auto` though :)

File: src/input/JoystickInputDriver.cc [hunk 6]
```diff
@@ -64,7 +64,9 @@ bool JoystickInputDriver::open()
 {
     stopRequest = false;
 
-    fd = ::open("/dev/input/js0", O_RDONLY);
+    char* path=const_cast<char*> (("/dev/input/js" + nr).c_str());
```
Comment (kintel) [inline src/input/JoystickInputDriver.cc]: This confuses me. Are there implementations of `open()` that doesn't take a `const char *`?
As we use `boost::format` other places, you could do that here as well; smth. like this:
```
auto path = boost::format("/dev/input/js %d") % nr;
this->fd = ::open(path.c_str(), O_RDONLY);
```

File: src/openscad.cc (hunk 5 above)
Comment (MichaelPFrey) [reply src/openscad.cc]: thank you, that is a mistake. I currently have not setup my build system to build the spaceNav driver.

File: src/input/SpaceNavInputDriver.cc [hunk 7]
```diff
@@ -115,8 +115,12 @@ void SpaceNavInputDriver::close()
 
 }
 
+void SpaceNavInputDriver::setDominantAxisOnly(bool var){
+    DominantAxisOnly = var;
```
Comment (kintel) [inline src/input/SpaceNavInputDriver.cc]: Capitalization typo, also use `this->` for clarity

File: src/input/JoystickInputDriver.cc (hunk 6 above)
Comment (MichaelPFrey) [reply src/input/JoystickInputDriver.cc]: thank you for your suggestion.
changed to boost::format

File: src/input/JoystickInputDriver.cc (hunk 1 above)
Comment (MichaelPFrey) [reply src/input/JoystickInputDriver.cc]: changed - thank you for your suggestion.

File: src/input/JoystickInputDriver.h (hunk 2 above)
Comment (MichaelPFrey) [reply src/input/JoystickInputDriver.h]: thank you - I did not know that syntax.

**Comment by t-paul** [comment]
Hmm, not sure if I made a mistake when merging, but it seems the driver specific enable/disable setting has no effect anymore.

**Comment by MichaelPFrey** [comment]
I will look into it. The thing that springs to my mind:
The driver specific enable/disable is mend to require a restart of openscad.
Drivers can not be enabled/disabled "on the fly".

**Comment by MichaelPFrey** [comment]
I found the issue. Fix is in progress.

**Pull request merged**